### PR TITLE
refactor: Extracting a base Driver struct

### PR DESCRIPTION
**Pull request by hairyhenderson**
This is mainly a de-duplication PR. I've extracted all of the common struct fields and put them into a new `DefaultDriver` struct.

There's been some mention of needing this before, for example in #1220 (and others).

I've tested this manually, but not extensively. There are a few other tweaks I want to make too, but I want to get feedback early to see if I'm on the right trail.

Signed-off-by: Dave Henderson <dhenderson@gmail.com>

### CODE REVIEW

**Comment by hairyhenderson** [comment]
(also, I'm breaking tests here - will follow up soon with a fix for that)

**Comment by nathanleclaire** [comment]
Ah, very nice, thanks -- looking like a good start -- I'll review shortly.

File: drivers/default.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	}
+}
+
+// GetSSHKeyPath -
+func (d *DefaultDriver) GetSSHKeyPath() string {
+	return d.GetStorePath("id_rsa")
+}
+
+// GetStorePath -
+func (d *DefaultDriver) GetStorePath(path string) string {
```
Comment (hairyhenderson) [inline drivers/default.go]: I don't like the name of this function, but I couldn't think of anything better at the time... Now I'm thinking something like `ResolveStorePath`... any thoughts?

File: commands/inspect_test.go [hunk 2]
```diff
@@ -29,7 +29,20 @@ func TestCmdInspectFormat(t *testing.T) {
...
+		MachineName    string
+		PrivateKeyPath string
+		SSHPort        int
+		SSHUser        string
+		SwarmDiscovery string
+		SwarmHost      string
+		SwarmMaster    bool
+		URL            string
+	}
+	expected, _ = json.MarshalIndent(&ExpectedDriver{MachineName: "test-a", URL: "unix:///var/run/docker.sock"}, "", "    ")
```
Comment (hairyhenderson) [inline commands/inspect_test.go]: This is what I get for changing the none driver...

**Comment by hairyhenderson** [comment]
Ok, I think I'm done messing around with this. I'm happy with it now.

**Comment by sthulb** [comment]
I like this :) We should have done it from the start!

**Comment by nathanleclaire** [comment]
When you get a chance, can you please squash it to one commit @hairyhenderson ?

**Comment by hairyhenderson** [comment]
@nathanleclaire - ok, rebased and squashed. PTAL :)

File: commands/inspect_test.go (hunk 2 above)
Comment (nathanleclaire) [inline commands/inspect_test.go]: Why swallow the error?

**Comment by nathanleclaire** [comment]
I'm getting errors attempting a create on DigitalOcean.

```console
Creating SSH key...
Creating Digital Ocean droplet...
Created droplet ID 5527853, IP address 45.55.150.61
Getting to WaitForSSH function...
Error getting IP address waiting for SSH: GetIP not implemented for unknown
Getting to WaitForSSH function...
Error getting IP address waiting for SSH: GetIP not implemented for unknown
```

It seems that it uses the default's `GetIP` function, and not the `GetIP` function from the `digitalocean.Driver`.

**Comment by hairyhenderson** [comment]
@nathanleclaire - err... weird. I'm pretty sure I tested with the DO driver. Maybe I screwed something up along the way. I'll dig into it.

**Comment by hairyhenderson** [comment]
@nathanleclaire - Ok, I can reproduce the issue. I'm not totally sure why it's happening with the DO driver, and not with the virtualbox driver...

**Comment by hairyhenderson** [comment]
Ok, this is happening because of my poor understanding of how polymorphism works in Go. The default implementation of `GetSSHHostname()` is:

```go
func (d *DefaultDriver) GetSSHHostname() (string, error) {
	return d.GetIP()
}
```

My _assumption_ was that the receiver here would magically change to the outer type, except of course it doesn't in Go. So I'm going to have to re-think this part.

File: commands/inspect_test.go (hunk 2 above)
Comment (hairyhenderson) [inline commands/inspect_test.go]: Yeah ok, handled now.

**Comment by hairyhenderson** [comment]
@nathanleclaire - ok, should be OK now. PTAL.

**Comment by nathanleclaire** [comment]
This is looking good...  I want to review and get in probably right after the release

**Comment by clnperez** [comment]
From how I understand this, the "Default Driver" doesn't work stand-alone, right? It seems like naming it "Default" kind of implies that it will work that way, and maybe do some looking around on your system and pick whatever environment you've got available (VirtualBox, qemu, etc.). So I'm thinking it could be named BaseDriver instead to reduce confusion (although users won't see the name of it at all, but, for developers new to the code).

**Comment by hairyhenderson** [comment]
@clnperez - right, `BaseDriver` is more descriptive. I'll rename it :)

**Comment by hairyhenderson** [comment]
@nathanleclaire @ehazlett - anything left to do before this can be merged?

**Comment by clnperez** [comment]
I was assuming you were going to rename the struct, etc. to BaseDriver instead of DefaultDriver. Maybe I misunderstood and you were just renaming the PR?

**Comment by hairyhenderson** [comment]
@clnperez - d'oh - I thought I did that but I guess I didn't... Sorry! Doing that now :)

**Comment by hairyhenderson** [comment]
Ok, renamed! :)

**Comment by clnperez** [comment]
Woot! Thanks @hairyhenderson

**Comment by nathanleclaire** [comment]
Intend to review this quite soon.

**Comment by nathanleclaire** [comment]
Nice work @hairyhenderson ... Testing looks pretty good on my end

LGTM

**Comment by ehazlett** [comment]
Yay!  LGTM

**Comment by hairyhenderson** [comment]
Thanks! 😁

**Pull request merged**